std/assembly/set.ts

```diff
@@ -108,6 +108,49 @@ export class Set<K> {
     }
   }
 
+  private getPtr(key: K): usize {
+    if (!this.has(key)) return EMPTY;
+    var oldPtr = changetype<usize>(this.entries) + HEADER_SIZE_AB;
+    var endPtr = oldPtr + <usize>this.entriesOffset * ENTRY_SIZE<K>();
+    while (oldPtr != endPtr) {
+      let entry = changetype<SetEntry<K>>(oldPtr);
+      if (!(entry.taggedNext & EMPTY) && entry.key == key) {
+        return oldPtr;
+      }
+      oldPtr += ENTRY_SIZE<K>();
+    }
+    return EMPTY;
+  }
+
+  forEach(callbackfn: (value1: K, value2: K, set: Set<K>) => void): void {
+    var startPtr = changetype<usize>(this.entries) + HEADER_SIZE_AB;
+    var endPtr = startPtr + <usize>this.entriesOffset * ENTRY_SIZE<K>();
+    var oldPtr = startPtr;
+    while (oldPtr != endPtr) {
+      let oldEntry = changetype<SetEntry<K>>(oldPtr);
+      oldPtr += ENTRY_SIZE<K>();
+      if (!(oldEntry.taggedNext & EMPTY)) {
+        callbackfn(oldEntry.key, oldEntry.key, this);
+        let currentStartPtr = changetype<usize>(this.entries) + HEADER_SIZE_AB;
+        // check if rehashing action triggered
+        if (startPtr != currentStartPtr) {
+          let newPtr = currentStartPtr;
+          do {
+            oldPtr -= ENTRY_SIZE<K>();
+            let visitedEntry = changetype<SetEntry<K>>(oldPtr);
+            if (!(visitedEntry.taggedNext & EMPTY) && this.has(visitedEntry.key)) {
+              newPtr = this.getPtr(visitedEntry.key) + ENTRY_SIZE<K>();
+              break;
+            }
+          } while (oldPtr != startPtr);
+          startPtr = currentStartPtr;
+          oldPtr = newPtr;
+        }
+      }
+      endPtr = startPtr + <usize>this.entriesOffset * ENTRY_SIZE<K>();
```

Comment:
Is it possible cache `<usize>this.entriesOffset * ENTRY_SIZE<K>()` to variable and move out of loop?

Reply:
Because of the rehash action  maybe trigger  or the end pointer maybe change in the callback function,  so can't move out of loop.

Reply:
I see